content/blog/2022/2022-02-14-spring-shell.md

```diff
@@ -0,0 +1,295 @@
+```java
+@ShellMethod(value = "add keys")  
+public void sshAdd(@ShellOption(value = "--k", arity = 2) String[] keys)  
+{  
+    log.info(format("Adding keys '%s' '%s'", keys[0], keys[1]));  
+}
+```
+
+Let's try the command out in the shell:
+
+```bash
+shell:>ssh-add --k test1 test2
+2022-02-12 18:27:00.301  INFO 4928 --- [           main] j.t.springshell.command.SSHCommand       : Adding keys 'test1' 'test2'
+```
+
+#### Boolean Parameters
+Boolean parameters receive a special treatment by command-line utilities. The absence of the parameter in the command indicates a false value. On the other hand, its existence indicates true value:
+```java
+@ShellMethod(value = "sign in")  
+public void sshLogin(@ShellOption(value = "--r") boolean rememberMe)  
+{  
+    log.info(format("remember me option is '%s'", rememberMe));  
+}
+```
+Let's check it out in the command line:
+```bash
+shell:>ssh-login --r
+2022-02-12 18:41:34.903  INFO 10044 --- [           main] j.t.springshell.command.SSHCommand       : remember me option is 'true'
+shell:>ssh-login
+2022-02-12 18:41:44.606  INFO 10044 --- [           main] j.t.springshell.command.SSHCommand       : remember me option is 'false'
+```
+#### Validating Command Parameters
+According to the Spring Shell team
+
+> Spring Shell integrates with the  [Bean Validation API](https://beanvalidation.org/)  to support automatic and self documenting constraints on command parameters.
+> Annotations found on command parameters as well as annotations at the
+> method level will be honored and trigger validation prior to the
+> command executing.
+
+Let's try this in action by adding a `@Size` annotation to the method parameter:
+```java
+@ShellMethod(value = "ssh agent")  
+public void sshAgent(  
+        @ShellOption(value = "--a")  
+        @Size(min = 2, max = 10) String agent)  
+{  
+    log.info(format("adding agent '%s'", agent));  
+}
+```
+Now, if we try to pass a parameter value with a length of 1 we will get an error stating the reason:
+```bash
+shell:>ssh-agent --a t
+The following constraints were not met:
+	--a string : size must be between 2 and 10 (You passed 't')
+```
+
+Note that the `@Size` annotation is a part of the [Jakarta Bean Validation](https://beanvalidation.org/) which offers many more validation options like `@NotEmpty` `@Max`.
+
+
+## Dynamic Command Availability
+Some commands only make sense when certain pre-conditions are met. For example, a `sign-out` command should be available only if a `sign-in` command has been issued, and if the user tries to run the `sign-out` command we want to warn them that it's not possible.
+
+Spring Shell offers us three ways to achieve our goal.
+
+Let's explore the **first way** which checks our class for a method with a special name and with a return type of `Availability`.
```

Comment:
The first, second and third way are a perfect opportunity to split up into sub sections to make it easier to read (and to make the table of contents more valuable)